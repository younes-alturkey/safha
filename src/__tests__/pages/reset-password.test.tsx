import '@testing-library/jest-dom'
import { cleanup, fireEvent, render, waitFor } from '@testing-library/react'
import { act } from 'react-dom/test-utils'
import TestWrapper from 'src/@core/components/test-wrapper'
import RestPasswordPage from 'src/pages/reset-password'

const pushMock = jest.fn()

jest.mock('next/router', () => ({
  useRouter: () => ({
    push: pushMock,
    events: {
      on: jest.fn(),
      off: jest.fn(),
      emit: jest.fn()
    },
    isFallback: false
  })
}))

jest.mock('next/navigation', () => ({
  useSearchParams: () => ({
    get: jest.fn()
  })
}))

jest.mock('src/api/auth', () => ({
  forgot: jest.fn().mockResolvedValue({ status: 200 })
}))

jest.mock('src/api/analytics', () => ({
  createEvent: jest.fn().mockResolvedValue({ status: 200, data: {} }),
  getIPInfo: jest.fn().mockResolvedValue({ status: 200, data: {} })
}))

jest.mock('next-auth/react', () => ({
  useSession: jest.fn().mockReturnValue({
    data: {
      user: {
        email: 'admin@safha.com'
      }
    }
  })
}))

let container: HTMLElement
let submitButton: HTMLButtonElement
let logoImage: HTMLImageElement
let emailInput: HTMLInputElement
const user = {
  email: 'testuser@example.com'
}
const redirectUrl = '/signin'

beforeEach(() => {
  jest.useFakeTimers()

  const rendered = render(
    <TestWrapper mode='light'>
      <RestPasswordPage apiUrl='https://sapi.safha.com' />
    </TestWrapper>
  )
  container = rendered.container
  submitButton = document.getElementById('submit-button') as HTMLButtonElement
  logoImage = document.getElementById('safha-logo') as HTMLImageElement
  emailInput = document.getElementById('email') as HTMLInputElement
})

afterEach(() => {
  jest.runOnlyPendingTimers()
  jest.useRealTimers()
  cleanup()
})

describe('Reset Password Page', () => {
  it('renders reset-password page unchanged', () => {
    expect(container).toMatchSnapshot()
  })

  it('displays the logo image', () => {
    expect(logoImage).toBeInTheDocument()
  })

  it('renders submit button correctly', () => {
    expect(submitButton).toBeInTheDocument()
  })

  it('renders submit button disabled', () => {
    expect(submitButton).toBeDisabled()
  })

  it('allows user to input email', async () => {
    await act(async () => {
      fireEvent.change(emailInput, { target: { value: user.email } })
    })

    expect(emailInput).toHaveValue(user.email)
  })

  it('sends forgot email successfully', async () => {
    await act(async () => {
      fireEvent.change(emailInput, { target: { value: user.email } })
      fireEvent.click(submitButton)
    })

    await waitFor(() => {
      expect(pushMock).toHaveBeenCalledWith(redirectUrl)
    })
  })

  it('handles form is reset after submission', async () => {
    await act(async () => {
      fireEvent.change(emailInput, { target: { value: user.email } })
      fireEvent.click(submitButton)
    })

    await waitFor(() => {
      expect(emailInput).toHaveValue('')
    })
  })
})
