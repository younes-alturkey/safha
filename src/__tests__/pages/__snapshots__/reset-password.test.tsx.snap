// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reset Password Page renders reset-password page unchanged 1`] = `
<div>
  <div
    class="layout-wrapper MuiBox-root css-33gw4"
  >
    <div
      class="app-content MuiBox-root css-10qwnza"
    >
      <div
        class="content-center MuiBox-root css-0"
      >
        <div
          class="MuiBox-root css-1i2bp0u"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation7 MuiCard-root css-tt7mc4-MuiPaper-root-MuiCard-root"
          >
            <div
              class="MuiCardContent-root css-10pa0r4-MuiCardContent-root"
            >
              <div
                class="MuiBox-root css-fdeg8e"
              >
                <img
                  alt="Safha Logo"
                  data-nimg="1"
                  data-testid="safha-logo"
                  decoding="async"
                  height="75"
                  id="safha-logo"
                  loading="lazy"
                  src="/_next/image?url=https%3A%2F%2Fstorage.googleapis.com%2Fsafha-public-01%2Fsafha-logo-transparent-black.png&w=384&q=75"
                  srcset="/_next/image?url=https%3A%2F%2Fstorage.googleapis.com%2Fsafha-public-01%2Fsafha-logo-transparent-black.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fstorage.googleapis.com%2Fsafha-public-01%2Fsafha-logo-transparent-black.png&w=384&q=75 2x"
                  style="color: transparent; opacity: 0.5;"
                  width="150"
                />
              </div>
              <form
                action="#"
                autocomplete="off"
                class="w-full"
                novalidate=""
              >
                <div
                  class="MuiBox-root css-79elbk"
                >
                  <div
                    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-11ns3xm-MuiFormControl-root-MuiTextField-root"
                    data-testid="email"
                  >
                    <label
                      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-filled MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-filled css-7peqn8-MuiFormLabel-root-MuiInputLabel-root"
                      data-shrink="true"
                      for="email"
                      id="email-label"
                    >
                      البريد الإلكتروني
                    </label>
                    <div
                      class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-ah3eer-MuiInputBase-root-MuiFilledInput-root"
                    >
                      <input
                        aria-invalid="false"
                        class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputSizeSmall css-8wyvbo-MuiInputBase-input-MuiFilledInput-input"
                        id="email"
                        placeholder="younes@safha.com"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                  <p
                    class="MuiTypography-root MuiTypography-body1 css-1msog3x-MuiTypography-root"
                  />
                </div>
                <button
                  class="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth Mui-disabled MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-y7ldic-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
                  data-testid="submit-button"
                  disabled=""
                  id="submit-button"
                  tabindex="-1"
                  type="submit"
                >
                  إعادة تعيين
                </button>
                <div
                  class="MuiBox-root css-z0xj7h"
                >
                  <a
                    class="MuiTypography-root MuiTypography-body1 css-1my6stc-MuiTypography-root"
                    href="/signin"
                  >
                    تسجيل الدخول
                  </a>
                </div>
              </form>
            </div>
            <button
              aria-label="switch app locale"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-hub4ho-MuiButtonBase-root-MuiIconButton-root"
              data-testid="switch_locale_button"
              id="switch_locale_button"
              tabindex="0"
              type="button"
            >
              <span />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              aria-label="switch app locale"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-hmdvb8-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <span />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
